libtenzir/builtins/operators/from_opensearch.cpp

```diff
@@ -0,0 +1,480 @@
+      .note("must be one of `brotli`, `bz2`, `gzip`, `lz4`, `zstd`")
+      .done();
+  }
+  auto out = std::vector<uint8_t>{};
+  out.resize(r.payload().size_bytes() * 2);
+  const auto codec = arrow::util::Codec::Create(
+    compression_type.ValueUnsafe(), arrow::util::kUseDefaultCompressionLevel);
+  TENZIR_ASSERT(codec.ok());
+  const auto decompressor
+    = codec.ValueUnsafe()->MakeDecompressor().ValueOrDie();
+  auto written = size_t{};
+  auto read = size_t{};
+  while (read != r.payload().size_bytes()) {
+    const auto result = decompressor->Decompress(
+      r.payload().size_bytes() - read,
+      reinterpret_cast<const uint8_t*>(r.payload().data() + read),
+      out.capacity() - written, out.data() + written);
+    if (not result.ok()) {
+      return diagnostic::warning("failed to decompress: {}",
+                                 result.status().ToString())
+        .done();
+    }
+    TENZIR_ASSERT(std::cmp_less_equal(result->bytes_written, out.size()));
+    written += result->bytes_written;
+    read += result->bytes_read;
+    if (result->need_more_output) {
+      if (out.size() == out.max_size()) [[unlikely]] {
+        return diagnostic::error("failed to resize buffer").done();
+      }
+      if (out.size() < out.max_size() / 2) {
+        out.resize(out.size() * 2);
+      } else [[unlikely]] {
+        out.resize(out.max_size());
+      }
+    }
+    // In case the input contains multiple concatenated compressed streams,
+    // we gracefully reset the decompressor.
+    if (decompressor->IsFinished()) {
+      const auto result = decompressor->Reset();
+      if (not result.ok()) {
+        return diagnostic::warning("failed to reset decompressor: {}",
+                                   result.ToString())
+          .done();
+      }
+    }
+  }
+  TENZIR_ASSERT(written != 0);
+  out.resize(written);
+  return chunk::make(std::move(out));
+}
+
+auto handle_slice(bool& is_action, table_slice slice) -> table_slice {
+  if (slice.rows() == 0) {
+    return {};
+  }
+  const auto ty = as<record_type>(slice.schema());
+  auto fields = std::vector<record_type::field_view>{};
+  for (const auto& field : ty.fields()) {
+    fields.push_back(field);
+  }
+  const auto delete_ = std::ranges::any_of(
+    fields,
+    [](auto&& x) {
+      return x == "delete";
+    },
+    &record_type::field_view::name);
+
+  const auto other_actions = std::ranges::any_of(
+    fields,
+    [](auto&& x) {
+      return x == "create" or x == "index" or x == "update";
+    },
+    &record_type::field_view::name);
+
+  if (delete_) {
+    return is_action ? table_slice{} : subslice(slice, 0, 1);
+  }
+  if (other_actions) {
+    auto filtered = std::vector<table_slice>{};
+    if (is_action) {
+      is_action = slice.rows() % 2 == 0;
+      for (auto i = size_t{1}; i < slice.rows(); i += 2) {
+        filtered.push_back(subslice(slice, i, i + 1));
+      }
+      return concatenate(filtered);
+    }
+    is_action = slice.rows() % 2 != 0;
```

Comment:
This heuristic is not very convincing. Have you considered inverting it and trying to detect properties of a data object instead?

Reply:
You cannot. An event can very much look the same as an action. E.g., if you get something that looks like an update action twice, then one of them is the action, and the other one is the event.